### PR TITLE
Create tags

### PR DESCRIPTION
**Pull request by bonifacyj**
This is code to create tags for both cluster and topic metadata

### CODE REVIEW

File: google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/scrape/metadata_scraper.py [hunk 1]
```diff
@@ -1,37 +1,117 @@
...
-        self._consumer = consumer
+    def __init__(self, client, bootstrap_server):
+        self._adminClient = client
+        self._bootstrap_server = bootstrap_server
 
     def get_metadata(self):
         try:
-            raw_metadata = self._consumer.list_topics(timeout=20)
-            topic_metadata = self._get_topic_metadata(raw_metadata)
+            raw_metadata = self._adminClient.list_topics(timeout=20)
```
Comment (mesmacosta) [inline google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/scrape/metadata_scraper.py]: use same naming convention: `_adminClient` > _admin_client

File: google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/scrape/metadata_scraper.py [hunk 2]
```diff
@@ -1,37 +1,117 @@
...
             MetadataConstants.CLUSTER_ID: cluster_id,
-            MetadataConstants.BROKERS_NUM: num_brokers
+            MetadataConstants.NUM_BROKERS: num_brokers,
+            MetadataConstants.BOOTSTRAP_SERVER: self._bootstrap_server
         }
         return cluster_metadata
+
+    def _get_topics_metadata(self, metadata_object):
+        topic_names = metadata_object.topics.keys()
+        descriptions = list()
```
Comment (mesmacosta) [inline google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/scrape/metadata_scraper.py]: initialize list with `[]`, so it's more concise?

File: google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/config/tag_template_constants.py [hunk 3]
```diff
@@ -24,6 +25,34 @@ def __init__(self, name, display_name, field_type, is_required=False):
...
+    class ClusterFields:
+
+        def __init__(self, num_brokers, num_topics, bootstrap_address):
+            self.num_brokers = num_brokers
+            self.num_topics = num_topics
+            self.bootstrap_address = bootstrap_address
+
+    class TopicFields:
+
+        def __init__(self, num_partitions, retention_time,
```
Comment (socha23) [inline google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/config/tag_template_constants.py]: should the fields be passed from the outside? Can we just create them in a (no-arg) constructor? Is there any case where we would want to pass different values based on some condition?

File: google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/scrape/metadata_scraper.py [hunk 4]
```diff
@@ -1,37 +1,117 @@
...
+        config_resources = [
+            ConfigResource(confluent_kafka.admin.RESOURCE_TOPIC, topic_name)
+            for topic_name in topic_names
+        ]
+        config_futures = self._adminClient.describe_configs(config_resources,
+                                                            request_timeout=10)
+        for topic, future in config_futures.items():
+            try:
+                config = future.result()
```
Comment (mesmacosta) [inline google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/scrape/metadata_scraper.py:52]: interested on how this behaves when testing this with 1000 topics.

Does it return 1 future for each topic?

I'd add some logging measuring this, so we can see if parallelizing the future fetching would be worth it.

File: google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/prepare/datacatalog_tag_template_factory_test.py [hunk 5]
```diff
@@ -21,7 +21,7 @@ def test_create_tag_template_for_cluster_metadata(self):
         self.assertEqual(tag_template.display_name,
                          "My_entry_group Cluster - Metadata")
         self._assert_fields(tag_template,
-                            TagTemplateConstants().field_constants_for_cluster)
+                            TagTemplateConstants().cluster_constants_list)
```
Comment (mesmacosta) [inline google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/prepare/datacatalog_tag_template_factory_test.py]: WDYT about the `TagTemplateConstants` not being an instance class?

Having to create a new instance of a Constant class each time you use it, feels that this class is doing more than simple storing constants.

So the suggestion would be, inside the `TagTemplateConstants`, remove the `__init__` and turn `_define_constants_for_cluster_tag_template`  and `_define_constants_for_topic_tag_template` to `@classmethod`, so you would be able to use it as `TagTemplateConstants.cluster_constants_list` and would be able to use the same instance.

**Review by mesmacosta — CHANGES_REQUESTED** [review]

File: google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/scrape/metadata_scraper.py (hunk 4 above)
Comment (bonifacyj) [reply google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/scrape/metadata_scraper.py:52]: According to the [doc](https://docs.confluent.io/current/clients/confluent-kafka-python/#confluent_kafka.admin.AdminClient), it does return one future for each topic. Could you elaborate on what you have in mind when talking about parallelizing the future fetching?

File: google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/prepare/datacatalog_tag_template_factory_test.py (hunk 5 above)
Comment (bonifacyj) [reply google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/prepare/datacatalog_tag_template_factory_test.py]: okay, I find it's a good idea.

File: google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/config/tag_template_constants.py (hunk 3 above)
Comment (bonifacyj) [reply google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/config/tag_template_constants.py]: not really, we can create them straight in the constructor. I think your approach actually creates more understandable outcome

File: google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/scrape/metadata_scraper.py (hunk 4 above)
Comment (mesmacosta) [reply google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/scrape/metadata_scraper.py:52]: The `future.result()` in the iteration loop is working as serial operations, being called for each topic, so each time it waits for the return to go to the next.

I'd add some logging to see how much time is spent there, so see if it is worth optimizing that.

You could combine batches of futures and parallelize that to fetch the results. But would only spent time evaluating this, if the amount of time is considerable. 

Since a Kafka cluster won't have that many topics to make it worth.

To sum up, for now I would suggest just adding the logging with the elapsed time spent on this.

Comment (bonifacyj) [reply google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/scrape/metadata_scraper.py:52]: Okay, I see! sure, let's log it.

Comment (bonifacyj) [reply google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/scrape/metadata_scraper.py:52]: I have measured it now, and description of 1000 topics takes approx 1.2 sec for me. This information is now also being logged in logging.info()

**Review by mesmacosta — APPROVED** [review]

File: google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/config/tag_template_constants.py [hunk 6]
```diff
@@ -10,60 +10,82 @@ class TagTemplateConstants:
     in the tag templates.
     """
 
-    def __init__(self):
-        self.field_constants_for_cluster = self.\
-            _define_constants_for_cluster_tag_template()
-        self.field_constants_for_topic = self.\
-            _define_constants_for_topic_tag_template()
+    @classmethod
+    def get_topic_constants_list(cls):
```
Comment (socha23) [inline google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/config/tag_template_constants.py:14]: can this be just 'get_topic_tag_fields' or something like that? I don't think the return type should be made explicit in the method name

File: google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/scrape/metadata_values_converter.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    '''
+    Duration-related values in Kafka, such as retention time,
+    are stored in ms, which might not always be human readable.
+    The same goes for space values, that Kafka stores in bytes.
+    This class converts those time and space values to
+    human-readable format.
+    '''
+
+    @staticmethod
+    def get_human_readable_duration_value(duration):
```
Comment (socha23) [inline google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/scrape/metadata_values_converter.py]: I'd make it explicit in the param name that this is milliseconds. `duration_ms` perhaps?

File: google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/scrape/metadata_values_converter_test.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+from google.datacatalog_connectors.kafka.scrape.metadata_values_converter \
+    import MetadataValuesConverter
+
+
+class MetadataValuesConverterTest(unittest.TestCase):
+
+    def test_duration_is_converted_to_human_redable_format(self):
+        input_in_ms = 31764000000
+        expected_output = '1 y 2 d 15 h 20 min'
```
Comment (socha23) [inline google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/scrape/metadata_values_converter_test.py]: Wonder if those spaces make it easier or harder to understand. `1 y 2 d 15 h 20 min` vs. `1y 2d 15h 20min`

File: google-datacatalog-kafka-connector/tools/cleanup_datacatalog.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        'At least one must be specified',
+        required=True)
+    parser.add_argument(
+        '--datacatalog-location-id',
+        help='Location id which is the Region that your Datacatalog resides',
+        default='us-central1')
+    return parser.parse_args()
+
+
+if __name__ == '__main__':
```
Comment (socha23) [inline google-datacatalog-kafka-connector/tools/cleanup_datacatalog.py:107]: Why do we need this? If it is used only in tests perhaps it should be in test directory, I'm not sure if we want to provide the user with a tool that deletes all their data

**Review by socha23 — APPROVED** [review]

File: google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/scrape/metadata_values_converter_test.py (hunk 8 above)
Comment (bonifacyj) [reply google-datacatalog-kafka-connector/tests/google/datacatalog_connectors/kafka/scrape/metadata_values_converter_test.py]: I think without spaces it's easier to read, and it's better to remove spaces

File: google-datacatalog-kafka-connector/tools/cleanup_datacatalog.py (hunk 9 above)
Comment (bonifacyj) [reply google-datacatalog-kafka-connector/tools/cleanup_datacatalog.py:107]: It's for developers - we assume that the connector is a sample code that could be extended or adapted based on users' needs. In README it is described how to use this cleanup.
Since all the connectors have similar structure and all of them have a cleanup script in tools/, I did the same here for Kafka.

File: google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/config/tag_template_constants.py (hunk 6 above)
Comment (bonifacyj) [reply google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/config/tag_template_constants.py:14]: hm, I wanted to explicitly separate this method from one delivering topic fields as an object (couple of lines below). Because in some places I need an object with attributes corresponding to tag template fields and in others - an iterable of those tag template fields. And to highlight the fact that it returns an iterable, I called it get_XYZ_list(). Maybe there are more pythonic ways to achieve that goal, wdyt?

Comment (socha23) [reply google-datacatalog-kafka-connector/src/google/datacatalog_connectors/kafka/config/tag_template_constants.py:14]: ah, so it makes sense to leave `list` here, to differentiate with the previous one. Let's leave it as it is.
